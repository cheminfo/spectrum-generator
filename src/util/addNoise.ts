import { randomUniform, randomNormal } from 'd3-random';
import XSAdd from 'ml-xsadd';

import type { AddNoiseOpt } from '../types/addNoiseOpt';
import type { Data } from '../types/data';

export default function addNoise(
  data: Data,
  percent = 0,
  options: AddNoiseOpt = {},
) {
  const { distribution = 'uniform', seed } = options;

  let generateRandomNumber;
  switch (distribution) {
    case 'uniform': {
      generateRandomNumber = getRandom(randomUniform, seed, -0.5, 0.5);
      break;
    }
    case 'normal': {
      generateRandomNumber = getRandom(randomNormal, seed);
      break;
    }
    default:
      throw new Error(`Unknown distribution ${distribution}`);
  }

  if (!percent) return data;
  let ys = data.y;
  let factor = (percent * findMax(ys)) / 100;
  for (let i = 0; i < ys.length; i++) {
    ys[i] += generateRandomNumber() * factor;
  }
  return data;
}

function getRandom(
  func: typeof randomNormal | typeof randomUniform,
  seed?: number,
  ...args: [number, number | undefined] | []
) {
  return typeof seed === 'number'
    ? func.source(new XSAdd(seed).random)(...args)
    : func(...args);
}

function findMax(array: Float64Array | number[]) {
  let max = Number.MIN_VALUE;
  for (let item of array) {
    if (item > max) max = item;
  }
  return max;
}
