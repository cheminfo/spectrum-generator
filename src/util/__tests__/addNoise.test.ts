import addNoise from '../addNoise';

test('addNoise', () => {
  let corrected = addNoise({ x: [1, 2, 3], y: [2, 3, 4] }, 10, { seed: 0 });
  expect(corrected.x).toHaveLength(3);
  expect(corrected).toMatchSnapshot();
});
