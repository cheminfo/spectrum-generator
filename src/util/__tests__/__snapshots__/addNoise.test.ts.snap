// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addNoise 1`] = `
Object {
  "x": Array [
    1,
    2,
    3,
  ],
  "y": Array [
    1.8588724613189698,
    2.914162039756775,
    4.052007699012757,
  ],
}
`;
